File: forge-core/internal/kerneldecisioncontract/closure.go
package kerneldecisioncontract

import (
	"crypto/sha256"
	"encoding/hex"
	"fmt"

	op "forgeos/forge-core/internal/kerneloperationalcontract"
)

func validateClosureBody(value *KernelDecisionReferenceClosure, blank bool) error {
	if value == nil {
		return fmt.Errorf("KernelDecisionReferenceClosure is nil")
	}
	if value.APIVersion != closureAPI || value.Canonicalization != canonicalization ||
		value.Kind != closureKind || value.Result != SuccessMarker {
		return fmt.Errorf("KernelDecisionReferenceClosure constants or marker differ")
	}
	if err := identity(value.ClosureID, value.ClosureSHA256, closurePrefix, "closure", blank); err != nil {
		return err
	}
	if err := validateAttestations(value.Attestations); err != nil {
		return err
	}
	if err := sortAtoms(value.CognitiveAtoms); err != nil {
		return err
	}
	if _, err := canonicalBytes(value.CognitiveAtoms, maxAtomSetBytes); err != nil {
		return fmt.Errorf("cognitive_atoms: %w", err)
	}
	if err := ValidateDecisionTransaction(&value.DecisionTransaction); err != nil {
		return err
	}
	if err := op.ValidateClosure(&value.OperationalClosure); err != nil {
		return fmt.Errorf("operational_closure: %w", err)
	}
	return validateReferenceGraph(value.CognitiveAtoms, &value.DecisionTransaction,
		&value.OperationalClosure)
}

func closureDigest(value *KernelDecisionReferenceClosure) (string, error) {
	blank := *value
	blank.ClosureID, blank.ClosureSHA256 = "", ""
	if err := validateClosureBody(&blank, true); err != nil {
		return "", err
	}
	raw, err := canonicalBytes(&blank, maxClosureBytes)
	if err != nil {
		return "", err
	}
	digest := sha256.New()
	digest.Write(closureDomain)
	digest.Write(raw)
	return hex.EncodeToString(digest.Sum(nil)), nil
}

func ValidateClosure(value *KernelDecisionReferenceClosure) error {
	if err := validateClosureBody(value, false); err != nil {
		return err
	}
	digest, err := closureDigest(value)
	if err != nil {
		return err
	}
	if value.ClosureSHA256 != digest {
		return fmt.Errorf("closure_sha256 does not match canonical preimage")
	}
	_, err = canonicalBytes(value, maxClosureBytes)
	return err
}

func SealClosure(value *KernelDecisionReferenceClosure) (*KernelDecisionReferenceClosure, error) {
	if value == nil || value.ClosureID != "" || value.ClosureSHA256 != "" {
		return nil, fmt.Errorf("sealing closure requires blank identity")
	}
	sealed, err := cloneValue(value)
	if err != nil {
		return nil, err
	}
	digest, err := closureDigest(sealed)
	if err != nil {
		return nil, err
	}
	sealed.ClosureID, sealed.ClosureSHA256 = closurePrefix+digest, digest
	return sealed, ValidateClosure(sealed)
}

func DecodeClosure(raw []byte) (*KernelDecisionReferenceClosure, error) {
	var value KernelDecisionReferenceClosure
	if err := decodeTyped(raw, maxClosureBytes, &value); err != nil {
		return nil, err
	}
	if _, err := op.DecodeClosure(mustOperationalBytes(&value.OperationalClosure)); err != nil {
		return nil, fmt.Errorf("operational_closure: %w", err)
	}
	return &value, ValidateClosure(&value)
}

func mustOperationalBytes(value *op.KernelOperationalReferenceClosure) []byte {
	raw, err := op.CanonicalJSON(value)
	if err != nil {
		return nil
	}
	return raw
}

func CanonicalJSON(value any) ([]byte, error) {
	return canonicalBytes(value, maxClosureBytes)
}

File: forge-runtime/crates/domain/src/kernel_decision_contract/closure.rs
use sha2::{Digest, Sha256};

use super::{
    CANONICALIZATION, CLOSURE_API, CLOSURE_DOMAIN, CLOSURE_KIND, CLOSURE_PREFIX,
    KernelDecisionContractError, KernelDecisionReferenceClosure, MAX_ATOM_SET_BYTES,
    MAX_CLOSURE_BYTES, SUCCESS_MARKER,
    atom::validate_atoms,
    graph::validate_reference_graph,
    invalid,
    primitives::{attestations, identity},
    transaction::validate_decision_transaction,
    wire,
};

fn validate_body(
    value: &KernelDecisionReferenceClosure,
    allow_blank: bool,
) -> Result<(), KernelDecisionContractError> {
    if value.api_version != CLOSURE_API
        || value.canonicalization != CANONICALIZATION
        || value.kind != CLOSURE_KIND
        || value.result != SUCCESS_MARKER
    {
        return Err(invalid(
            "KernelDecisionReferenceClosure constants or marker differ",
        ));
    }
    identity(
        &value.closure_id,
        &value.closure_sha256,
        CLOSURE_PREFIX,
        "closure",
        allow_blank,
    )?;
    attestations(&value.attestations)?;
    validate_atoms(&value.cognitive_atoms)?;
    wire::canonical_with_max(&value.cognitive_atoms, MAX_ATOM_SET_BYTES)?;
    validate_decision_transaction(&value.decision_transaction)?;
    crate::kernel_operational_contract::validate_closure(&value.operational_closure)
        .map_err(|error| invalid(format!("operational_closure: {}", error.message)))?;
    validate_reference_graph(
        &value.cognitive_atoms,
        &value.decision_transaction,
        &value.operational_closure,
    )
}

fn digest(value: &KernelDecisionReferenceClosure) -> Result<String, KernelDecisionContractError> {
    let mut blank = value.clone();
    blank.closure_id.clear();
    blank.closure_sha256.clear();
    validate_body(&blank, true)?;
    let canonical = wire::canonical_with_max(&blank, MAX_CLOSURE_BYTES)?;
    let mut hasher = Sha256::new();
    hasher.update(CLOSURE_DOMAIN);
    hasher.update(canonical.as_bytes());
    Ok(crate::governance_contract::codec::lower_hex(
        &hasher.finalize(),
    ))
}

/// Validates one exact sealed Kernel decision reference closure and complete one-way DAG.
///
/// # Errors
///
/// Returns an error for any atom, transaction, reused operational closure, graph, or digest violation.
pub fn validate_closure(
    value: &KernelDecisionReferenceClosure,
) -> Result<(), KernelDecisionContractError> {
    validate_body(value, false)?;
    if value.closure_sha256 != digest(value)? {
        return Err(invalid("closure_sha256 does not match canonical preimage"));
    }
    wire::canonical_with_max(value, MAX_CLOSURE_BYTES)?;
    Ok(())
}

/// Seals one exact blank-identity Kernel decision reference closure copy.
///
/// # Errors
///
/// Returns an error for nonblank identity or any invalid nested record, graph edge, or preimage.
pub fn seal_closure(
    value: &KernelDecisionReferenceClosure,
) -> Result<KernelDecisionReferenceClosure, KernelDecisionContractError> {
    if !value.closure_id.is_empty() || !value.closure_sha256.is_empty() {
        return Err(invalid("sealing closure requires blank identity"));
    }
    let mut sealed = value.clone();
    let digest = digest(&sealed)?;
    sealed.closure_id = format!("{CLOSURE_PREFIX}{digest}");
    sealed.closure_sha256 = digest;
    validate_closure(&sealed)?;
    Ok(sealed)
}

/// Decodes exact compact canonical Kernel decision reference closure bytes.
///
/// # Errors
///
/// Returns an error for malformed, noncanonical, oversized, invalid, unresolved, or unsealed bytes.
pub fn decode_closure(
    bytes: &[u8],
) -> Result<KernelDecisionReferenceClosure, KernelDecisionContractError> {
    let value: KernelDecisionReferenceClosure = wire::decode_typed(bytes, MAX_CLOSURE_BYTES)?;
    let operational =
        crate::kernel_operational_contract::canonical_json(&value.operational_closure)
            .map_err(|error| invalid(format!("operational_closure encode: {}", error.message)))?;
    crate::kernel_operational_contract::decode_closure(operational.as_bytes())
        .map_err(|error| invalid(format!("operational_closure: {}", error.message)))?;
    validate_closure(&value)?;
    Ok(value)
}
